pkg/act/act.go

```diff
@@ -0,0 +1,295 @@
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/metrics/metric"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/service"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+const (
+	metricsUpdateInterval = 15 * time.Second
+	metricsTimeout        = 10 * time.Minute
+)
+
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "act-metrics")
+
+var Cell = cell.Module(
+	"act-metrics",
+	"Metrics with counts of new and active connections for each service-zone pair",
+
+	metrics.Metric(NewActiveConnectionTrackingMetrics),
+	cell.Provide(NewACT),
+	cell.Invoke(func(*ACT) {}),
+)
+
+type ActiveConnectionTrackingMetrics struct {
+	New, Active, Failed metric.DeletableVec[metric.Gauge]
+	ProcessingTime      metric.Histogram
+}
+
+func NewActiveConnectionTrackingMetrics() ActiveConnectionTrackingMetrics {
+	return ActiveConnectionTrackingMetrics{
+		New: metric.NewGaugeVec(metric.GaugeOpts{
+			ConfigName: metrics.Namespace + "_act_new_connections_total",
+			Subsystem:  "act",
+			Namespace:  metrics.Namespace,
+			Name:       "new_connections_total",
+		}, []string{"zone", "service"}),
+		Active: metric.NewGaugeVec(metric.GaugeOpts{
+			ConfigName: metrics.Namespace + "_act_active_connections_total",
+			Subsystem:  "act",
+			Namespace:  metrics.Namespace,
+			Name:       "active_connections_total",
+		}, []string{"zone", "service"}),
+		Failed: metric.NewGaugeVec(metric.GaugeOpts{
+			ConfigName: metrics.Namespace + "_act_failed_connections_total",
+			Subsystem:  "act",
+			Namespace:  metrics.Namespace,
+			Name:       "failed_connections_total",
+			Help:       "number of service connections purged from conntrack table",
+		}, []string{"zone", "service"}),
+		ProcessingTime: metric.NewHistogram(metric.HistogramOpts{
+			ConfigName: metrics.Namespace + "_act_processing_time_seconds",
+			Subsystem:  "act",
+			Namespace:  metrics.Namespace,
+			Name:       "processing_time_seconds",
+			Help:       "time to go over ACT map and update the metrics",
+		}),
+	}
+}
+
+type actMetric struct {
+	opened, closed    uint64
+	newFailed, failed uint64
+	updated           time.Time
+	labelValues       []string
+}
+
+type ACT struct {
+	src          act.ActiveConnectionTrackingMap
+	metrics      ActiveConnectionTrackingMetrics
+	keyToStrings func(key *act.ActiveConnectionTrackerKey) (zone string, svc string, err error)
+
+	// mux protects tracker map
+	mux *lock.Mutex
+	// tracker is a map[zone][svc]metric
+	tracker map[uint8]map[uint16]*actMetric
+}
+
+func NewACT(in struct {
+	cell.In
+
+	Lifecycle cell.Lifecycle
+	Source    act.ActiveConnectionTrackingMap
+	Metrics   ActiveConnectionTrackingMetrics
+}) *ACT {
+	if in.Source == nil {
+		// Active Connection Tracking is disabled.
+		return nil
+	}
+	a := newAct(in.Source, in.Metrics, option.Config)
+	ctx, cancel := context.WithCancel(context.Background())
+	in.Lifecycle.Append(cell.Hook{
+		OnStart: func(_ cell.HookContext) error {
+			go a.run(ctx)
```

Comment:
Let's use https://pkg.go.dev/github.com/cilium/hive/job instead of a bare goroutine. You can depend on a job.Group in the parameter struct, and it is automatically added to the lifecycle and scoped to the module. It also gives you access to `cell.Health` to report health of the component, as needed.